File: scripts/reference-solutions/three-sum-closest.py
import sys

data = list(map(int, sys.stdin.read().strip().split()))
n, target = data[0], data[1]
nums = sorted(data[2 : 2 + n])

best = nums[0] + nums[1] + nums[2]

for i in range(n - 2):
    left = i + 1
    right = n - 1
    while left < right:
        total = nums[i] + nums[left] + nums[right]
        if abs(total - target) < abs(best - target) or (
            abs(total - target) == abs(best - target) and total < best
        ):
            best = total

        if total < target:
            left += 1
        elif total > target:
            right -= 1
        else:
            print(total)
            sys.exit(0)

print(best)

File: scripts/reference-solutions/three-sum-closest.cpp
#include <algorithm>
#include <cstdlib>
#include <iostream>
#include <vector>

using namespace std;

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  int n;
  long long target;
  cin >> n >> target;

  vector<long long> nums(n);
  for (long long& value : nums) {
    cin >> value;
  }

  sort(nums.begin(), nums.end());
  long long best = nums[0] + nums[1] + nums[2];

  for (int i = 0; i < n - 2; ++i) {
    int left = i + 1;
    int right = n - 1;
    while (left < right) {
      long long total = nums[i] + nums[left] + nums[right];
      long long totalDistance = llabs(total - target);
      long long bestDistance = llabs(best - target);
      if (totalDistance < bestDistance || (totalDistance == bestDistance && total < best)) {
        best = total;
      }

      if (total < target) {
        ++left;
      } else if (total > target) {
        --right;
      } else {
        cout << total << '\n';
        return 0;
      }
    }
  }

  cout << best << '\n';
  return 0;
}
